ports/esp32/machine_can.c

```diff
@@ -0,0 +1,673 @@
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_alert_obj, machine_hw_can_alert);
+
+// Clear TX Queue
+STATIC mp_obj_t machine_hw_can_clear_tx_queue(mp_obj_t self_in){
+    return mp_obj_new_bool(can_clear_transmit_queue() == ESP_OK);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_clear_tx_queue_obj, machine_hw_can_clear_tx_queue);
+
+// Clear RX Queue
+STATIC mp_obj_t machine_hw_can_clear_rx_queue(mp_obj_t self_in){
+    return mp_obj_new_bool(can_clear_receive_queue() == ESP_OK);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_clear_rx_queue_obj, machine_hw_can_clear_rx_queue);
+
+// send([data], id, *)
+STATIC mp_obj_t machine_hw_can_send(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args){
+    enum{
+        ARG_data,
+        ARG_id,
+        ARG_timeout,
+        ARG_rtr,
+        ARG_self
+    };
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_data, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL}},
+        {MP_QSTR_id, MP_ARG_REQUIRED | MP_ARG_INT, {.u_int = 0}},
+        {MP_QSTR_timeout, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0}},
+        {MP_QSTR_rtr, MP_ARG_KW_ONLY | MP_ARG_BOOL, {.u_bool = false}},
+    };
+
+    // parse args
+    machine_can_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // populate message
+    size_t length;
+    mp_obj_t *items;
+    mp_obj_get_array(args[ARG_data].u_obj, &length, &items);
+    if (length > 8){
+        mp_raise_ValueError("CAN data field too long");
+    }
+    uint8_t flags = (args[ARG_rtr].u_bool == true ? CAN_MSG_FLAG_RTR : CAN_MSG_FLAG_NONE);
+    uint32_t id = args[ARG_id].u_int;
+    if (self->extframe){
+        flags += CAN_MSG_FLAG_EXTD;
+        id &= 0x1FFFFFFF;
+    }else{
+        id &= 0x1FF;
+    }
+    if (self->loopback){
+        flags += CAN_MSG_FLAG_SELF;
+    }
+    can_message_t tx_msg = {.data_length_code = length,
+                            .identifier = id,
+                            .flags = flags};
+    for (uint8_t i = 0; i < length; i++){
+        tx_msg.data[i] = mp_obj_get_int(items[i]);
+    }
+    if (_machine_hw_can_get_status().state == CAN_STATE_RUNNING){
+        int status = can_transmit(&tx_msg, args[ARG_timeout].u_int);
+        if (status != ESP_OK){
+            mp_raise_OSError(-status);
+        }
+        return mp_const_none;
+    }else{
+        nlr_raise(mp_obj_new_exception_msg(&mp_type_RuntimeError, "CAN Device is not ready"));
+    }
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(machine_hw_can_send_obj, 3, machine_hw_can_send);
+
+// recv(list=None, *, timeout=5000)
+STATIC mp_obj_t machine_hw_can_recv(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args){
+    enum{
+        ARG_list,
+        ARG_timeout
+    };
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_list, MP_ARG_OBJ, {.u_rom_obj = MP_ROM_PTR(&mp_const_none_obj)}},
```

Comment:
`.u_rom_obj = MP_ROM_NONE`

And spaces around the `{` and `}`. e.g.  `{ MP_QSTR_list, MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} }` seems to be the convention.

Reply:
compiler cannot resold MP_ROM_NONE, I have to add any further header file?

Reply:
I think this definition is not available since my branch started before this addition. Actually I don't know I rebase the branch without making a mess.

Reply:
The branch should rebase cleanly (I checked). You'll need to do this anyway, and clean up the commits a bit in the process. Probably the best way to do this is with an interactive rebase.

```
git fetch origin
git rebase -i origin/master
```

You'll be prompted to select what you want to do with each of the commits. I suggest using "fix" to combine commits together (and possibly reorder them). And then possibly run the `rebase -i` again and go through and use `reword` to update the commit messages.